src/Storages/System/StorageSystemDetachedParts.cpp

```diff
@@ -8,108 +8,208 @@
 {
-    /// Files or directories of detached part may not exist. Only count the size of existing files.
     if (disk->isFile(from))
     {
-        total_size += disk->getFileSize(from);
+        files.push_back(from);
     }
     else
     {
         for (auto it = disk->iterateDirectory(from); it->isValid(); it->next())
-            calculateTotalSizeOnDiskImpl(disk, fs::path(from) / it->name(), total_size);
+            partFilesOnDiskImpl(disk, fs::path(from) / it->name(), files);
     }
 }
 
-static UInt64 calculateTotalSizeOnDisk(const DiskPtr & disk, const String & from)
+std::vector<String> partFilesOnDisk(const DiskPtr & disk, const String & from)
 {
-    UInt64 total_size = 0;
+    std::vector<String> files;
     try
     {
-        calculateTotalSizeOnDiskImpl(disk, from, total_size);
+        partFilesOnDiskImpl(disk, from, files);
     }
     catch (...)
     {
         tryLogCurrentException(__PRETTY_FUNCTION__);
     }
-    return total_size;
+    return files;
 }
 
-Pipe StorageSystemDetachedParts::read(
-    const Names & column_names,
-    const StorageSnapshotPtr & storage_snapshot,
-    SelectQueryInfo & query_info,
-    ContextPtr context,
-    QueryProcessingStage::Enum /*processed_stage*/,
-    const size_t /*max_block_size*/,
-    const size_t /*num_streams*/)
+class SourceState
 {
-    storage_snapshot->check(column_names);
+    std::mutex mutex;
+    StoragesInfoStream stream;
 
-    StoragesInfoStream stream(query_info, context);
+public:
+    SourceState(StoragesInfoStream && stream_)
+        : stream(std::move(stream_))
+    {}
 
-    /// Create the result.
-    Block block = storage_snapshot->metadata->getSampleBlock();
+    StoragesInfo next()
+    {
+        std::lock_guard lock(mutex);
+        return stream.next();
+    }
+};
 
-    NameSet names_set(column_names.begin(), column_names.end());
-    std::vector<UInt8> columns_mask(block.columns());
-    Block header;
+class DetachedPartsSource : public ISource
+{
+public:
+    DetachedPartsSource(Block header_, std::shared_ptr<SourceState> state_, std::vector<UInt8> columns_mask_, UInt64 block_size_,
+                        bool has_bytes_on_disk_column_)
+        : ISource(std::move(header_))
+        , state(state_)
+        , columns_mask(std::move(columns_mask_))
+        , block_size(block_size_)
+        , has_bytes_on_disk_column(has_bytes_on_disk_column_)
+    {}
 
-    for (size_t i = 0; i < block.columns(); ++i)
+    String getName() const override { return "DataPartsSource"; }
+
+protected:
+    Chunk generate() override
     {
-        if (names_set.contains(block.getByPosition(i).name))
+        Chunk result;
+
+        while (result.getNumRows() < block_size)
         {
-            columns_mask[i] = 1;
-            header.insert(block.getByPosition(i));
+            if (detached_parts.empty())
+                getMoreParts();
+
+            if (detached_parts.empty())
+            {
+                progress(result.getNumRows(), result.bytes());
```

Comment:
We don't need to call progress here, and it would be called automatically. It is only necessary to call progress if there is some internal filtration.